File: Assignment3/src/main/java/org/example/structural/entity/Book.java
package org.example.structural.entity;

import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.example.structural.service.BookDecorator;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Book implements BookDecorator {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    Long id;
    String title;
    String author;
    double price;
    String category;

    public String getDescription() {
        return title + " by " + author + (category != null ? " (" + category + ")" : "");
    }

    @Override
    public double getPrice() {
        return price;
    }
}

File: Assignment3/frontend/lib/domain/entities/book.dart
class Book {
  final int? id; // Nullable for creation, present for retrieval
  final String category;
  final String title;
  final String author;
  final double price;

  const Book({
    this.id,
    required this.category,
    required this.title,
    required this.author,
    required this.price,
  });
}